### PR TITLE
`sage --fixdoctests --update-known-test-failures`; silence modularized distributions in CI

### PR DESCRIPTION
**Pull request by mkoeppe**
The distributions `sagemath-repl` and `sagemath-categories` run parts of the testsuite in a virtual environment created by `tox` when SAGE_CHECK=yes. This also runs as part of the Build & Test CI.

Here we make several improvements:
- Support SAGE_CHECK=warn
- Invoke the doctester with `--baseline-stats-path={toxinidir}/known-test-failures.json`
- Make the tests silent unless new test failures, not recorded in the `known-test-failures.json`, are encountered

To help maintain the `known-test-failures.json`, we also add new features to the command `sage --fixdoctests`.
- `sage --fixdoctests --update-known-test-failures` reads the stats files generated by the doctester in the virtual environments and writes updated `known-test-failures.json` files to the source tree
- doctester stats now also include a field `ntests` - number of doctests of a module that were run

`sage --fixdoctests` also receives new switches `--distribution all`, `--fixed-point`, `--verbose`, `--no-diff` and some other improvements.

- Part of: #29705 
- Cherry-picked from: #35095 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

- Depends on #36263 (merged here)
- Depends on #36238 (merged here)

### CODE REVIEW

File: pkgs/sagemath-environment/tox.ini [hunk 1]
```diff
@@ -85,14 +85,15 @@ package_env = .pkg-sagepython
...
 [testenv:sagepython-sagewheels]
 basepython = {env:SAGE_VENV}/bin/python
 package_env = .pkg-sagepython
 
 [testenv:sagepython-norequirements]
 basepython = {env:SAGE_VENV}/bin/python3
 package_env = .pkg-sagepython
+
+
+[testenv:sagepython-sagewheels-nopypi-norequirements]
```
Comment (kwankyu) [inline pkgs/sagemath-environment/tox.ini:97]: single empty line

File: pkgs/sagemath-objects/tox.ini [hunk 2]
```diff
@@ -89,14 +89,15 @@ package_env = .pkg-sagepython
...
 [testenv:sagepython-sagewheels]
 basepython = {env:SAGE_VENV}/bin/python
 package_env = .pkg-sagepython
 
 [testenv:sagepython-norequirements]
 basepython = {env:SAGE_VENV}/bin/python3
 package_env = .pkg-sagepython
+
+
+[testenv:sagepython-sagewheels-nopypi-norequirements]
```
Comment (kwankyu) [inline pkgs/sagemath-objects/tox.ini:101]: Two lines, intended?

File: pkgs/sagemath-repl/known-test-failures.json [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        "failed": true,
+        "ntests": 28
+    },
+    "sage.misc.sage_input": {
+        "failed": true,
+        "ntests": 721
+    },
+    "sage.repl.attach": {
+        "ntests": 128
+    },
```
Comment (kwankyu) [inline pkgs/sagemath-repl/known-test-failures.json:46]: No "failed: false"?

Or "success: true"?

File: pkgs/sagemath-standard/tox.ini [hunk 4]
```diff
@@ -16,23 +16,20 @@
...
     #
-    sagepython-sagewheels-nopypi-norequirements,
+    sagepython-sagewheels-nopypi-norequirements
+    #
+    # OTHER SUPPORTED ENVIRONMENTS:
+    #
+    # Build dependencies according to requirements.txt (all versions fixed).
+    # Use ONLY the wheels built and stored by the Sage distribution (no PyPI):
+    #
+    #   ./sage -sh -c '(cd pkgs/sagemath-standard && tox -v -v -v -e sagepython-sagewheels-nopypi)'
```
Comment (kwankyu) [inline pkgs/sagemath-standard/tox.ini:32]: No `sagepython-sagewheels-nopypi` here?

File: pkgs/sagemath-standard/tox.ini [hunk 5]
```diff
@@ -16,23 +16,20 @@
...
-    #
-    # SUPPORTED ENVIRONMENTS:
-    #
-    # Build dependencies according to requirements.txt (all versions fixed).
-    # Use ONLY the wheels built and stored by the Sage distribution (no PyPI):
-    #
-    #   ./sage -sh -c '(cd pkgs/sagemath-standard && tox -v -v -v -e sagepython-sagewheels-nopypi)'
-    #
-    sagepython-sagewheels-nopypi,
-    #
```
Comment (kwankyu) [inline pkgs/sagemath-standard/tox.ini:28]:
```
    #
    # SUPPORTED ENVIRONMENTS:
    #
```
No?

File: pkgs/sagemath-environment/tox.ini (hunk 1 above)
Comment (mkoeppe) [reply pkgs/sagemath-environment/tox.ini:97]: This break is intended - this sets apart the default environment

File: pkgs/sagemath-repl/known-test-failures.json (hunk 3 above)
Comment (mkoeppe) [reply pkgs/sagemath-repl/known-test-failures.json:46]: No, the timings2.json files created by the doctester don't have these.

File: pkgs/sagemath-standard/tox.ini (hunk 4 above)
Comment (mkoeppe) [reply pkgs/sagemath-standard/tox.ini:32]: No, the `envlist` only lists the default environments, i.e., those that will be tested if one uses `tox` without the `-e` switch.

File: pkgs/sagemath-standard/tox.ini (hunk 5 above)
Comment (mkoeppe) [reply pkgs/sagemath-standard/tox.ini:28]: I've changed the format from "SUPPORTED" / "OTHER SUPPORTED" to (unmarked) default / "OTHER SUPPORTED"

File: pkgs/sagemath-objects/tox.ini (hunk 2 above)
Comment (mkoeppe) [reply pkgs/sagemath-objects/tox.ini:101]: Yes, this separates the default environment from the others above

**Comment by kwankyu** [comment]
I got 
```
$ sage --fixdoctests
sage-fixdoctests: At least one filename is required when --update-known-test-failures is not used
Running "sage -t -p "
```
Is it really running "sage -t -p "?

**Comment by kwankyu** [comment]
```
 $ sage --fixdoctests --help
...
  --probe FEATURES      check whether '# optional/needs' tags are still
                        needed, remove these
...
```
`remove those not needed`?

**Comment by kwankyu** [comment]
From 
```
$ sage --fixdoctests --update-known-test-failures --verbose
Running "sage -t -p "
```
I don't understand the output. It seems not running anything.

**Comment by kwankyu** [comment]
"--fixed-point" sounds good by itself. But since it follows "--fixdoctests", it makes me stop and think. An alternative is "--stabilize". Just a suggestion.

**Comment by kwankyu** [comment]
```
$ sage --fixdoctests --distribution sagemath-objects --verbose src/sage/rings/function_field
Running "pkgs/sagemath-objects/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage -t --environment sage.all__sagemath_objects ./src/sage/version.py"
sh: pkgs/sagemath-objects/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage: No such file or directory
sage-fixdoctests: Doctester exited with error status 127
```
Is this normal? I ran this after `make wheels`.

**Comment by kwankyu** [comment]
```
$ sage --fixdoctests --distribution sagemath-categories --verbose src/sage/rings/function_field --fixed-point
##############################################################################
sage-fixdoctests: Fixing with --distribution=sagemath-categories
Running "pkgs/sagemath-categories/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage -t --environment sage.all__sagemath_categories ./src/sage/version.py"
Running "pkgs/sagemath-categories/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage -t -p --environment sage.all__sagemath_categories --if-installed src/sage/rings/function_field/all.py src/sage/rings/function_field/constructor.py src/sage/rings/function_field/derivations.py src/sage/rings/function_field/derivations_polymod.py 
...
src/sage/rings/function_field/valuation_ring.py"
usage: sage -t [options] filenames
sage-runtests: error: unrecognized arguments: --if-installed
sage-fixdoctests: Processing unprocessed files
sage-fixdoctests: No fixes made in 'src/sage/rings/function_field/place.py'
...
sage-fixdoctests: Fixed point reached
```
There is an error `sage-runtests: error: unrecognized arguments: --if-installed` above.

**Comment by kwankyu** [comment]
From 
```
$ sage --fixdoctests --distribution sagemath-categories --verbose src/sage/rings/function_field/function_field.py --fixed-point
##############################################################################
sage-fixdoctests: Fixing with --distribution=sagemath-categories
Running "pkgs/sagemath-categories/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage -t --environment sage.all__sagemath_categories ./src/sage/version.py"
Running "pkgs/sagemath-categories/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage -t -p --environment sage.all__sagemath_categories --if-installed src/sage/rings/function_field/function_field.py"
usage: sage -t [options] filenames
sage-runtests: error: unrecognized arguments: --if-installed
sage-fixdoctests: Processing unprocessed files
sage-fixdoctests: No fixes made in 'src/sage/rings/function_field/function_field.py'
sage-fixdoctests: Fixed point reached
```
why does this run?
```
Running "pkgs/sagemath-categories/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage -t 
  --environment sage.all__sagemath_categories ./src/sage/version.py"
```
It seems that this runs if `--distribution` is given.

**Comment by mkoeppe** [comment]
> ```
> $ sage --fixdoctests
> sage-fixdoctests: At least one filename is required when --update-known-test-failures is not used
> Running "sage -t -p "
> ```
> 
> Is it really running "sage -t -p "?

Thanks, fixed in 93f6a05d197

**Comment by mkoeppe** [comment]
> ```
> $ sage --fixdoctests --help
> ...
>   --probe FEATURES      check whether '# optional/needs' tags are still
>                         needed, remove these
> ...
> ```
> 
> 
>     
>   
> 
> `remove those not needed`?

Thanks, done in e380f9a3024

**Comment by mkoeppe** [comment]
> ```
> $ sage --fixdoctests --update-known-test-failures --verbose
> Running "sage -t -p "
> ```
> 
> 
>     
>   
> 
> I don't understand the output. It seems not running anything.

It gives an error now after a61cdbf7131

**Comment by mkoeppe** [comment]
> ```
> $ sage --fixdoctests --distribution sagemath-objects --verbose src/sage/rings/function_field
> Running "pkgs/sagemath-objects/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage -t --environment sage.all__sagemath_objects ./src/sage/version.py"
> sh: pkgs/sagemath-objects/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage: No such file or directory
> sage-fixdoctests: Doctester exited with error status 127
> ```
> 
> 
>     
>   
> 
> Is this normal? I ran this after `make wheels`.

The `.tox` environments are only created if you run that with `SAGE_CHECK=yes` (or `SAGE_CHECK=warn`)

**Comment by mkoeppe** [comment]
> There is an error `sage-runtests: error: unrecognized arguments: --if-installed` above.

This must have been an old `.tox` environment

**Comment by mkoeppe** [comment]
> why does this run?
> 
> ```
> Running "pkgs/sagemath-categories/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage -t 
>   --environment sage.all__sagemath_categories ./src/sage/version.py"
> ```
> 
> It seems that this runs if `--distribution` is given.

Before it tests the given files, it uses this command to see if the `.tox` environment is working. It skips those that do not seem to be working

**Comment by mkoeppe** [comment]
> "--fixed-point" sounds good by itself. But since it follows "--fixdoctests", it makes me stop and think. An alternative is "--stabilize". Just a suggestion.

I also considered `--until-stable` but ultimately I thought that "stable" might be taken as promising something good, whereas "fixed point" expresses that it has converged to "something", which is closer to the truth

**Comment by kwankyu** [comment]
> > why does this run?
> > ```
> > Running "pkgs/sagemath-categories/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage -t 
> >   --environment sage.all__sagemath_categories ./src/sage/version.py"
> > ```
> >   
> > It seems that this runs if `--distribution` is given.
> 
> Before it tests the given files, it uses this command to see if the `.tox` environment is working. It skips those that do not seem to be working

Then how about not logging the line "Running ... ./src/sage/version.py", but instead logging the result of the check? That log looks internal.

**Comment by kwankyu** [comment]
> > ```
> > $ sage --fixdoctests --distribution sagemath-objects --verbose src/sage/rings/function_field
> > Running "pkgs/sagemath-objects/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage -t --environment 
> > Is this normal? I ran this after `make wheels`.
> 
> The `.tox` environments are only created if you run that with `SAGE_CHECK=yes` (or `SAGE_CHECK=warn`)

The `.tox` environment is there, but `sage` is not in `bin`.

**Comment by mkoeppe** [comment]
> So the problem is with `sagemath_objects`

Ah, yes. The doctester is provided by **sagemath-repl**, which depends on **sagemath-objects**. Because of this, **sagemath-objects** is not tested using the doctester. (Also, the `sage` script is provided by **sagemath-environment**, which is also not installed in the tox environment of **sagemath-objects**.)

**Comment by kwankyu** [comment]
I suspected that sort of things. Then how about detecting such a situation, and warn the user of `--fixdoctests`. If `--distribution all` is used, then `sagemath-objects` should be skipped anyway instead of erroring out. Right?

**Comment by mkoeppe** [comment]
> > Before it tests the given files, it uses this command to see if the `.tox` environment is working. It skips those that do not seem to be working
> 
> Then how about not logging the line "Running ... ./src/sage/version.py", but instead logging the result of the check? That log looks internal.

I've changed the wording in e3c4b15; better?

**Comment by mkoeppe** [comment]
> If `--distribution all` is used, then `sagemath-objects` should be skipped anyway

That's right, in #35095 'all' also expands to a list that does not include **sagemath-objects**.

**Comment by kwankyu** [comment]
```
$ sage --fixdoctests --distribution all src/sage/rings/function_field/function_field.py --fixed-point
##############################################################################
sage-fixdoctests: Fixing with --distribution=sagemath-categories
Running "pkgs/sagemath-categories/.tox/sagepython-sagewheels-nopypi-norequirements/bin/sage -t -p --environment sage.all__sagemath_categories --if-installed src/sage/rings/function_field/function_field.py"
sage-runtests: Skipping 'src/sage/rings/function_field/function_field.py' because module sage.rings.function_field cannot be imported
##############################################################################
sage-fixdoctests: Fixing with --distribution=''
Running "sage -t -p src/sage/rings/function_field/function_field.py"
sage-fixdoctests: Processing unprocessed files
sage-fixdoctests: No fixes made in 'src/sage/rings/function_field/function_field.py'
sage-fixdoctests: Fixed point reached
```
Why only the "empty distributon" and sagemath-categories?

**Comment by kwankyu** [comment]
Is the list of distributions hard-wired?

**Comment by mkoeppe** [comment]
Yes, it's hardwired; these are the only distributions for which the development branch has a useful testsuite.

Much more in #35095

**Comment by kwankyu** [comment]
I see. OK.

**Comment by kwankyu** [comment]
I will look through the code once again.

Otherwise, it looks good to me.

File: pkgs/sagemath-environment/tox.ini (hunk 1 above)
Comment (kwankyu) [reply pkgs/sagemath-environment/tox.ini:97]: OK.

File: pkgs/sagemath-objects/tox.ini (hunk 2 above)
Comment (kwankyu) [reply pkgs/sagemath-objects/tox.ini:101]: OK.

File: pkgs/sagemath-standard/tox.ini (hunk 4 above)
Comment (kwankyu) [reply pkgs/sagemath-standard/tox.ini:32]: OK. Thanks.

File: src/doc/en/developer/doctesting.rst [hunk 6]
```diff
@@ -1350,6 +1353,22 @@ utilized most efficiently)::
...
+The recorded pass/fail status of the files can be used for running only those files
+that failed their most recent test by using the ``--failed`` flag (``-f`` for short).
+
+Using the option ``--baseline-stats-path known-test-failures.json``,
+it is possible to distinguish files with known doctest failures
+from new failures. The file ``known-test-failures.json`` should be
+prepared in the same format as ``timings2.json``.
+
+Source files that marked as failed there will be marked as "[failed in baseline]"
+failures in the doctest report; and if there are only baseline failures, no
```
Comment (kwankyu) [inline src/doc/en/developer/doctesting.rst:1368]: `Source files marked as failed there...` or `Source files that are marked as failed there`?

File: src/doc/en/developer/doctesting.rst [hunk 7]
```diff
@@ -1655,3 +1677,23 @@ to the doctest.
...
+
+The modularized distribution packages ``pkgs/sagemath-categories`` and
+``pkgs/sagemath-repl`` contain files ``known-test-failures*.json`` for use
+with the option ``--baseline-stats-path``, see section
+:ref:`section-doctest-auxiliary-files`.
+
+These files can be updated using the command::
+
+    [mkoeppe@localhost sage]$ ./sage --fixdoctests --no-test                        \
+                                --update-known-test-failures --distribution all
```
Comment (kwankyu) [inline src/doc/en/developer/doctesting.rst:1702]: Is it right that `--no-test` here?

Comment (kwankyu) [reply src/doc/en/developer/doctesting.rst:1702]: `--update-known-test-failures` could work without doctesting?

File: src/doc/en/developer/doctesting.rst (hunk 6 above)
Comment (mkoeppe) [reply src/doc/en/developer/doctesting.rst:1368]: Fixed in 970f1df593e

File: src/doc/en/developer/doctesting.rst (hunk 7 above)
Comment (mkoeppe) [reply src/doc/en/developer/doctesting.rst:1702]: Yes, without `--no-test`, it will run the doctesters first. (Not sure if that's very useful.)

Comment (kwankyu) [reply src/doc/en/developer/doctesting.rst:1702]: `--update-known-test-failures --no-test` uses the information from previous doctests run?

Comment (kwankyu) [reply src/doc/en/developer/doctesting.rst:1702]: Perhaps from `timings2.json`?

Comment (mkoeppe) [reply src/doc/en/developer/doctesting.rst:1702]: Yes; it's always coming from timings2.json

Comment (kwankyu) [reply src/doc/en/developer/doctesting.rst:1702]: If so, it would not hurt to write a line that says so following the above lines...

Comment (kwankyu) [reply src/doc/en/developer/doctesting.rst:1702]: Or I may be misunderstanding...

Comment (mkoeppe) [reply src/doc/en/developer/doctesting.rst:1702]: I've made a change in 6aca56a; is this going in the right direction?

Comment (kwankyu) [reply src/doc/en/developer/doctesting.rst:1702]: Yes. That is very clear. Thanks.

**Comment by kwankyu** [comment]
There are genuine doctest failures like
```
File "src/sage/doctest/reporting.py", line 266, in sage.doctest.reporting.DocTestReporter.report
Failed example:
    DTR.stats
Expected:
    {'sage.doctest.reporting': {'failed': True, 'walltime': 1000000.0}}
Got:
    {'sage.doctest.reporting': {'failed': True,
      'ntests': 0,
      'walltime': 1000000.0}}
***************************************
```

**Comment by kwankyu** [comment]
I am not sure if the other failures are unrelated. As we are touching doctester, some files may be affected somehow...

**Comment by mkoeppe** [comment]
Indeed. The change in `src/sage/doctest/forker.py` is to blame. I'll drop the commit that made the change; I'll revisit this topic (of managing the global environment) in #35095

**Review by kwankyu — APPROVED** [review]
LGTM.

**Comment by mkoeppe** [comment]
Thank you!

**Pull request merged**